File: C++/996.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
    void print(const vector<int>& vec) {
        for (int a : vec) cout << a << ' ';
        cout << '\n';
    }
    int N;
    int count;
    unordered_map<int, int> B;
    unordered_map<int, vector<int> > squares;
    vector<int> vec;
    void test() {
        //print(vec);
        if (vec.size() == N) { count++; return; }
        for (int cand : squares[vec.back()]) {
            if (B[cand] == 0) continue;
            B[cand]--;
            vec.push_back(cand);
            test();
            vec.pop_back();
            B[cand]++;
        }
    }
    bool is_square(const int a) {
        const int r = (int)sqrt(a);
        return a == r * r;
    }
public:
    int numSquarefulPerms(vector<int>& A) {
        N = A.size();
        count = 0;
        for (const int& a : A) B[a]++;
        for (auto it : B) {
            for (auto it_ : B) {
                if (it.first <= it_.first) continue;
                if (is_square(it.first + it_.first)) {
                    //cout << "find pair " << it.first << " and " << it_.first << '\n';
                    squares[it.first].push_back(it_.first);
                    squares[it_.first].push_back(it.first);
                }
            }
            if (is_square(it.first * 2)) {
                //cout << "find pair " << it.first << " and " << it.first << '\n';
                squares[it.first].push_back(it.first);
            }
        }
        for (auto it : B) {
            B[it.first]--;
            vec.push_back(it.first);
            test();
            vec.pop_back();
            B[it.first]++;
        }
        return count;
    }
};
__________________________________________________________________________________________________
sample 8528 kb submission
class Solution {
public:
    int numSquarefulPerms(vector<int> &A) {
        preProcess(A);
        dfs(A, 0, 0);
        return result;
    }

    void dfs(vector<int> &A, int depth, int curr) {
        if (depth == A.size()) {
            result++; return;
        }
        int same = -1;
        for (int i = 0; i < A.size(); i++) {
            if (depth > 0 && !isSquare[curr][i]) {
                continue;
            }
            if (A[i] != same && !isVisited[i]) {
                isVisited[i] = 1;
                dfs(A, depth + 1, i);
                isVisited[i] = 0;
                same = A[i];
            }
        }
    }


private:
    static const int MAXN = 14;
    int isSquare[MAXN][MAXN];
    int isVisited[MAXN];
    int result;

    void preProcess(vector<int> &A) {
        sort(A.begin(), A.end());
        memset(isSquare, 0x00, MAXN * MAXN * sizeof(int));
        for (int i = 0; i < A.size(); i++) {
            for (int j = 0; j < i; j++) {
                int y = A[i] + A[j];
                int x = int(sqrt(double(y)));
                if (x * x == y) {
                    isSquare[i][j] = 1;
                    isSquare[j][i] = 1;
                }
            }
        }
        memset(isVisited, 0x00, MAXN * sizeof(int));
        result = 0;
    }
};
__________________________________________________________________________________________________

File: Python3/996.py
__________________________________________________________________________________________________
sample 28 ms submission
import math
class Solution:
    def numSquarefulPerms(self, A: List[int]) -> int:
        squares = {}
        left = {}
        empty_set = set()
        
        ret = [0]

        for i in range(len(A)):
            for j in range(i + 1, len(A)):
                sqrt = int(math.sqrt(A[i] + A[j]))
                
                if sqrt * sqrt == (A[i] + A[j]):
                    s = squares.setdefault(A[i], set())
                    s.add(A[j])
                    s = squares.setdefault(A[j], set())
                    s.add(A[i])
                    
            if A[i] not in left:
                left[A[i]] = 0
            left[A[i]] += 1
                    
        def dfs(prev, i):
            if i == len(A):
                ret[0] += 1
                return
            
            for val in list(left.keys()):
                
                if val in squares.get(prev, set()):
                    if (left[val] == 1):
                        del left[val]
                    else:
                        left[val] -= 1
                    dfs(val, i + 1)
                    if val not in left:
                        left[val] = 0
                    left[val] += 1
        
        
        for val in list(left.keys()):
            if (left[val] == 1):
                del left[val]
            else:
                left[val] -= 1
            dfs(val, 1)
            if val not in left:
                left[val] = 0
            left[val] += 1
            
        return ret[0]
__________________________________________________________________________________________________
sample 13136 kb submission
class Solution:
    def numSquarefulPerms(self, A: List[int]) -> int:
        count = collections.Counter(A)
        comb = {i: {j for j in count if int((i+j)**0.5)**2 == (i+j)} for i in count}
        self.result = 0
        def dfs(x, deep):
            if deep == 0:
                self.result += 1
                return
            count[x] -= 1
            for i in comb[x]:
                if count[i]:
                    dfs(i, deep - 1)
            count[x] += 1
        for i in comb:
            dfs(i, len(A)-1)
        return self.result
__________________________________________________________________________________________________
